internal/zinc-core/src/main/scala/sbt/internal/inc/MappedVirtualFile.scala

```diff
@@ -0,0 +1,85 @@
+/*
+ * Zinc - The incremental compiler for Scala.
+ * Copyright Lightbend, Inc. and Mark Harrah
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package sbt
+package internal
+package inc
+
+import java.io.InputStream
+import java.nio.file.{ Files, Path, Paths }
+import xsbti.{ BasicVirtualFileRef, FileConverter, PathBasedFile, VirtualFile, VirtualFileRef }
+
+class MappedVirtualFile(encodedPath: String, rootPathsMap: Map[Int, Path])
+    extends BasicVirtualFileRef(encodedPath)
+    with PathBasedFile {
+  private def path: Path = {
+    rootPathsMap.toSeq.find({
+      case (idx, p) =>
+        encodedPath.startsWith("${" + idx.toString + "}/")
+    }) match {
+      case Some((idx, p)) =>
+        p.resolve(encodedPath.drop(("${" + idx.toString + "}/").size))
+      case None =>
+        Paths.get(encodedPath)
+    }
+  }
+  override def contentHash: Long = HashUtil.farmHash(path)
+  override def input(): InputStream = Files.newInputStream(path)
+  def toPath(): Path = path
+}
+
+object MappedVirtualFile {
+  def apply(encodedPath: String, rootPathsMap: Map[Int, Path]): MappedVirtualFile =
+    new MappedVirtualFile(encodedPath, rootPathsMap)
+}
+
+class MappedFileConverter(rootPaths: Seq[Path], allowMachinePath: Boolean) extends FileConverter {
+  val rootPathIdx = rootPaths.zipWithIndex.map({ case (p, idx) => idx -> p })
+  val rootPathIdxUri = rootPathIdx flatMap {
+    case (idx, p) =>
+      val u = p.toUri.toString
+      if (u.startsWith("file:///var/folders/"))
+        Seq(
+          idx -> u,
+          idx -> u.replaceAllLiterally("file:///var/folders/", "file:///private/var/folders/")
+        )
```

Comment:
What is this special case for? It's missing documentation

Reply:
On macOS `/tmp/` and `/var/` seems to be symlinks to `/private/tmp/` and `/private/var/` respectively.

```
lrwxr-xr-x@   1 root  wheel    20xx tmp@ -> private/tmp
lrwxr-xr-x@   1 root  wheel    20xx var@ -> private/var
```

When temporary directories are created under `/var/folders/` Zinc and compilers sometimes end up reporting different paths for `*.scala`, `*.java`, `*.class` etc. Maybe there's a better solution than this, but the above is a workaround for that.

Reply:
I suggest using `getCanonicalPath`.

Reply:
(or `Path#toRealPath` which seems to be equivalent for Path, though this one will throw an exception if the file doesn't actually exist on disk)

Reply:
Yea. I have to figure out the right point to dealias symlinks so that a build can be created under `/tmp/` (or compiler bridge can be compiled in temp etc) and things continue to work.